modules/apps/29-fee/keeper/escrow.go

```diff
@@ -44,66 +44,143 @@ func (k Keeper) EscrowPacketFee(ctx sdk.Context, identifiedFee *types.Identified
-	if err != nil {
-		return sdkerrors.Wrap(err, "failed to send fee to forward relayer")
+	// cache context before trying to send to forward relayer
+	cacheCtx, writeFn := ctx.CacheContext()
+
+	fwd, err := sdk.AccAddressFromBech32(forwardRelayer)
+	if err == nil {
+		err = k.bankKeeper.SendCoinsFromModuleToAccount(cacheCtx, types.ModuleName, fwd, feeInEscrow.Fee.ReceiveFee)
+		if err == nil {
+			// write the cache
+			writeFn()
+			// NOTE: The context returned by CacheContext() refers to a new EventManager, so it needs to explicitly set events to the original context.
+			ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
+		}
+	} else {
+		distributeFwdErr = sdkerrors.Wrap(err, "failed to send fee to forward relayer")
 	}
 
-	// send ack fee to reverse relayer
-	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, reverseRelayer, feeInEscrow.Fee.AckFee)
-	if err != nil {
-		return sdkerrors.Wrap(err, "error sending fee to reverse relayer")
+	// cache context before trying to send to reverse relayer
+	cacheCtx, writeFn = ctx.CacheContext()
+
+	reverse, err := sdk.AccAddressFromBech32(reverseRelayer)
+	if err == nil {
+		err = k.bankKeeper.SendCoinsFromModuleToAccount(cacheCtx, types.ModuleName, reverse, feeInEscrow.Fee.AckFee)
+		if err == nil {
+			// write the cache
+			writeFn()
+			// NOTE: The context returned by CacheContext() refers to a new EventManager, so it needs to explicitly set events to the original context.
+			ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
+		}
+	} else {
+		distributeReverseErr = sdkerrors.Wrap(err, "failed to send fee to reverse relayer")
 	}
 
-	// refund timeout fee to refundAddr
-	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, refundAcc, feeInEscrow.Fee.TimeoutFee)
-	if err != nil {
-		return sdkerrors.Wrap(err, "error refunding timeout fee")
+	// cache context before trying to send timeout fee to refund address
+	cacheCtx, writeFn = ctx.CacheContext()
+
+	refund, err := sdk.AccAddressFromBech32(refundAcc)
+	if err == nil {
+		err = k.bankKeeper.SendCoinsFromModuleToAccount(cacheCtx, types.ModuleName, refund, feeInEscrow.Fee.TimeoutFee)
+		if err == nil {
+			// write the cache
+			writeFn()
+			// NOTE: The context returned by CacheContext() refers to a new EventManager, so it needs to explicitly set events to the original context.
+			ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
+		}
+	} else {
+		refundAccErr = sdkerrors.Wrap(err, "refunding timeout fee")
 	}
 
 	// removes the fee from the store as fee is now paid
 	k.DeleteFeeInEscrow(ctx, packetID)
 
+	// check if there are any errors, otherwise return nil
+	if packetIdErr != nil || distributeFwdErr != nil || distributeReverseErr != nil || refundAccErr != nil {
+		return sdkerrors.Wrapf(types.ErrFeeDistribution, "error in distributing fee: %s %s %s %s", packetIdErr, distributeFwdErr, distributeReverseErr, refundAccErr)
```

Comment:
no error should be returned (as this would cause the tx to revert)

the error can be emitted in an event (or perhaps logged)